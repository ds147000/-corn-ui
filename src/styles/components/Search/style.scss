.corn-search {
  height: 100px;
  width: 100%;
  box-sizing: border-box;
  padding: 24px 24px 12px;

  &-back {
    font-size: 24px;
    margin-right: 32px;
  }

  &-input-box {
    flex: 1;
    height: 62px;
    border-radius: 31px;
    box-sizing: border-box;
    position: relative;
    // padding: 17px 24px 17px 64px;
    overflow: hidden;
    background-color: $searchDefualtColor;
  }

  &-input {
    display: block;
    padding-left: 80px;
    height: 62px;
    font-size: 28px;
    font-weight: 400;
    color: $text1;
    line-height: 62px;
    flex: 1;
  }

  &-icon {
    position: absolute;
    font-size: 24px;
    color: $icon2;
    top: 19px;
    left: 24px;
    line-height: 24px;
  }

  &-clear {
    font-size: 24px;
    line-height: 24px;
    color: $icon2;
    height: 62px;
    width: 62px;
    z-index: 10;
    margin-right: 10px;
  }

  &-btn {
    height: 62px;
    font-weight: 400;
    width: 116px;
    margin-left: 0;
  }

  &-suffix {
    padding-left: 34px;
  }

  &-light {
    background-color: rgba($color: #fff , $alpha: 1);

    .corn-search-input-box {
      background-color: $searchLightColor;
    }
  }

  &-placeholder {
    width: 450px;
    height: 28px;
    overflow: hidden;
    pointer-events: none;
    font-size: 28px;
    font-weight: 400;
    color: $text3;
    line-height: 28px;
    position: absolute;
    z-index: 2;
    left: 64px;
    top: 17px;

    &-list {
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      transition: top 0.5s;
    }

    &-item {
      height: 30px;
    }
  }
}
