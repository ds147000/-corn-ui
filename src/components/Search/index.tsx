import React, { useCallback, useMemo, useRef, useState } from 'react'
import { View } from '@tarojs/components'
import classNames from 'classnames'
import Icon from '../Icon'
import Form from '../Form'
import Input from '../Input'
import Button from '../Button'
import { BaseEventOrig, ITouchEvent } from '../../types'
import { Input as InputType } from '../../types/Input'
import { SearchPlaceholder } from './placeholder'

export interface SearchProps {
  /** 风格类型*/
  type?: 'default' | 'light'
  /** 启用返回按钮 */
  back?: boolean
  /** 尾部插入的节点内容 */
  suffix?: React.ReactNode
  /** 可以点击清除图标删除内容,是否开启 */
  allowClear?: boolean
  /** 可输入 */
  openInput?: boolean
  /** 提示占位语, 可以字符串和数组字符串。如果是传入字符串数组将启用滚动提示语动画 */
  placeholder?: string | string[]
  /** 搜索值 */
  value?: string
  /** 点击输入框事件 */
  onClick?(event: ITouchEvent): void
  /** 输入发生改变事件 */
  onChange?(value: string): void
  /** 返回事件 */
  onBack?(): void
  /** 搜索事件 */
  onSearch?(value: string): void
}

const Search: React.FC<SearchProps> = ({
  type,
  back,
  openInput,
  suffix,
  value,
  placeholder,
  allowClear = true,
  onChange,
  onSearch,
  onBack,
  onClick
}) => {
  const [ isFouce, setIsFouce ] = useState(false)
  const [ isClear, setIsClear ] = useState(false)
  const searchForm = useRef<Form>()
  const _class = useMemo(() => classNames('xrk-search xrk-f xrk-ac', `xrk-search-${type}`), [ type ])

  const _onInput = (event: BaseEventOrig<InputType.inputEventDetail>): void => {
    onChange?.(event.detail.value)
    if (allowClear) setIsClear(true)
  }

  const _onSubmit = (event: Record<string, unknown>): void => {
    const searchValue = event.value as string
    onSearch?.(searchValue)
  }

  const _onClear = (): void => {
    console.log('情况')
    searchForm.current?.reset()
    setIsClear(false)
    setIsFouce(false)
    onChange?.('')
  }

  const _onFouce = useCallback(() => setIsFouce(true), [])
  const _onBlur = useCallback((event: BaseEventOrig<InputType.inputValueEventDetail>) => {
    if (!event.detail.value) setIsFouce(false)
  }, [])

  return (
    <Form onSubmit={_onSubmit} ref={searchForm as React.LegacyRef<Form>} >
      <View className={_class} >
        {back && <Icon className="xrk-search-back" name="left" data-testid="back" onClick={onBack} />}

        <View
          className="xrk-search-input-box"
          data-testid="search-box"
          onClick={onClick}
        >
          <Icon name="search" className="xrk-search-icon" />
          {!isFouce && Boolean(value) === false && <SearchPlaceholder data={placeholder} />}

          {openInput ? (
            <Input
              name="value"
              className="xrk-search-input"
              value={value}
              onInput={_onInput}
              onFocus={_onFouce}
              onBlur={_onBlur}
              autoComplete="off"
            />
          ) : (
            <View className="xrk-search-input" />
          )}

          {allowClear && isClear && <Icon name="clear" className="xrk-search-clear" onClick={_onClear} />}

          <Button
            size="middle"
            className="xrk-search-btn"
            data-testid="search-btn"
            formType="submit"
          >
            搜索
          </Button>
        </View>
        {Boolean(suffix) && <View className="xrk-search-suffix xrk-f xkr-ac" >{suffix}</View>}
      </View>
    </Form>
  )
}

export default Search
