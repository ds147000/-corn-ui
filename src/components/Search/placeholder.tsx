import React, { useEffect, useRef, useState } from 'react'
import { View } from '@tarojs/components'
import { transformRem } from '../../utils'

/** 不执行动画的长度 */
const STOP_ANIMTED_LEN = 1
/** 关键词展示时长 */
const ANIMTED_YIME = 5000
/** 初始化移动值 */
const DEFAULT_MOVE = 0
/** 移动速度 */
const MOVE_SEEPD = -30
/** 移动自增 */
const MOVE_ADD = 1
/** 移动最大值 */
const MOVE_MAX = 2

interface SearchPlaceholderListProps {
  data: string[]
}

const SearchPlaceholderList: React.FC<SearchPlaceholderListProps> = ({ data }) => {
  const [ move, setMove ] = useState(DEFAULT_MOVE)
  const time = useRef<NodeJS.Timeout>()

  useEffect(() => {
    if (data.length <= STOP_ANIMTED_LEN) return undefined

    function animtedMove(): void {
      time.current = setTimeout(() => {
        setMove((oldState): number => {
          if (oldState > data.length - MOVE_MAX) return DEFAULT_MOVE
          return oldState += MOVE_ADD
        })
        animtedMove()
      }, ANIMTED_YIME)
    }
    animtedMove()

    return (): void => {
      clearTimeout(time.current as NodeJS.Timeout)
    }
  }, [ data.length ])

  return (
    <View className="corn-search-placeholder" >
      <View className="corn-search-placeholder-list" style={{ top: transformRem(move * MOVE_SEEPD) }} >
        {data.map((item, key) => (
          <View key={item + String(key)} className="corn-search-placeholder-item corn-f corn-ac" >{item}</View>
        ))}
      </View>
    </View>
  )
}


export interface SearchPlaceholderProps {
  data?: string | string[]
}

export const SearchPlaceholder: React.FC<SearchPlaceholderProps> = ({ data }) => {

  if (!data) return null

  if (typeof data === 'string' || typeof data === 'number') {
    return (
      <View className="corn-search-placeholder">
        {data}
      </View>
    )
  }

  if (Array.isArray(data)) return <SearchPlaceholderList data={data} />

  return null
}

