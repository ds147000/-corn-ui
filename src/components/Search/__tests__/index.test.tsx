import React from 'react'
import { fireEvent, render, waitFor, createEvent } from '@testing-library/react'
import UserEvent from '@testing-library/user-event'
import Search from '../index'
import { act } from 'react-dom/test-utils'

jest.mock('../../Form')

describe('Search', () => {
  test('基础快照', async () => {
    const screen = render(<Search />)
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('基础快照 light', async () => {
    const screen = render(<Search type="light" />)
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('插入尾部内容', async () => {
    const onCLick = jest.fn()
    const screen = render(<Search suffix={<div onClick={onCLick} >打卡</div>} />)
    fireEvent.click(screen.getByText('打卡'))
    await waitFor(() => expect(screen.container).toMatchSnapshot())
    await waitFor(() => expect(onCLick).toHaveBeenCalledTimes(1))
  })

  test('开启返回', async () => {
    const onBack = jest.fn()
    const screen = render(<Search back onBack={onBack} />)
    fireEvent.click(screen.getByTestId('back'))
    await waitFor(() => expect(onBack).toHaveBeenCalledTimes(1))
  })

  test('关闭返回', async () => {
    const onBack = jest.fn()
    const screen = render(<Search back={false} onBack={onBack} />)
    await waitFor(async () => expect(await screen.queryByTestId('back')).toBeNull())
    await waitFor(() => expect(onBack).toHaveBeenCalledTimes(0))
  })

  test('输入值，搜索', async () => {
    const onSearch = jest.fn()
    const onChange = jest.fn()
    const screen = render(<Search openInput onSearch={onSearch} onChange={onChange} />)
    const input = screen.getByTestId('input')
    UserEvent.type(input, '喜马拉雅山')
    UserEvent.click(screen.getByTestId('search-btn'))
    await waitFor(() => expect(onChange).toHaveBeenCalledTimes(5))
    await waitFor(() => expect(onSearch).toHaveBeenCalledTimes(1))
    await waitFor(() => expect(onChange.mock.calls[0][0]).toBe('喜'))
    await waitFor(() => expect(onChange.mock.calls[1][0]).toBe('喜马'))
    await waitFor(() => expect(onChange.mock.calls[2][0]).toBe('喜马拉'))
    await waitFor(() => expect(onChange.mock.calls[3][0]).toBe('喜马拉雅'))
    await waitFor(() => expect(onChange.mock.calls[4][0]).toBe('喜马拉雅山'))
    await waitFor(() => expect(onSearch.mock.calls[0][0]).toBe('喜马拉雅山'))
  })


  test('输入值，搜索 无onChange监听', async () => {
    const onSearch = jest.fn()
    const screen = render(<Search openInput onSearch={onSearch}  />)
    const input = screen.getByTestId('input')
    UserEvent.type(input, '喜马拉雅山1234')
    UserEvent.click(screen.getByTestId('search-btn'))
    await waitFor(() => expect(onSearch).toHaveBeenCalledTimes(1))
    await waitFor(() => expect(onSearch.mock.calls[0][0]).toBe('喜马拉雅山1234'))
  })

  test('输入值，搜索 无事件情况下', async () => {
    const screen = render(<Search openInput />)
    const input = screen.getByTestId('input')
    UserEvent.type(input, '喜马拉雅山')
    UserEvent.click(screen.getByTestId('search-btn'))
  })

  test('输入值，点击清空按钮。但是配置不可清空  -> 搜索', async () => {
    const onSearch = jest.fn()
    const onChange = jest.fn()
    const screen = render(<Search openInput allowClear={false} onSearch={onSearch} onChange={onChange} />)
    const input = screen.getByTestId('input')
    UserEvent.type(input, '喜马拉雅山')
    UserEvent.click(screen.getByTestId('search-btn'))
    await waitFor(() => expect(screen.container.querySelector('.xrk-search-clear')).toBeNull())
    await waitFor(() => expect(onChange).toHaveBeenCalledTimes(5))
    await waitFor(() => expect(onSearch).toHaveBeenCalledTimes(1))
    await waitFor(() => expect(onChange.mock.calls[0][0]).toBe('喜'))
    await waitFor(() => expect(onChange.mock.calls[1][0]).toBe('喜马'))
    await waitFor(() => expect(onChange.mock.calls[2][0]).toBe('喜马拉'))
    await waitFor(() => expect(onChange.mock.calls[3][0]).toBe('喜马拉雅'))
    await waitFor(() => expect(onChange.mock.calls[4][0]).toBe('喜马拉雅山'))
    await waitFor(() => expect(onSearch.mock.calls[0][0]).toBe('喜马拉雅山'))
  })

  test('空值，搜索', async () => {
    const onSearch = jest.fn()
    const onChange = jest.fn()
    const screen = render(<Search openInput onSearch={onSearch} onChange={onChange} />)
    act(() => {
      UserEvent.click(screen.getByTestId('search-btn'))
    })
    await waitFor(() => expect(onChange).toHaveBeenCalledTimes(0))
    await waitFor(() => expect(onSearch).toHaveBeenCalledTimes(1))
    await waitFor(() => expect(onSearch.mock.calls[0][0]).toBe(''))
  })

  test('输入值，清空。 搜索', async () => {
    const onSearch = jest.fn()
    const onChange = jest.fn()
    const screen = render(<Search onSearch={onSearch} openInput onChange={onChange} allowClear />)
    const input = screen.getByTestId('input')
    UserEvent.type(input, '喜马拉雅山')
    UserEvent.click(screen.container.querySelector('.xrk-search-clear'))
    UserEvent.click(screen.getByTestId('search-btn'))
    await waitFor(() => expect(onChange).toHaveBeenCalledTimes(6))
    await waitFor(() => expect(onSearch).toHaveBeenCalledTimes(1))
    await waitFor(() => expect(onChange.mock.calls[0][0]).toBe('喜'))
    await waitFor(() => expect(onChange.mock.calls[1][0]).toBe('喜马'))
    await waitFor(() => expect(onChange.mock.calls[2][0]).toBe('喜马拉'))
    await waitFor(() => expect(onChange.mock.calls[3][0]).toBe('喜马拉雅'))
    await waitFor(() => expect(onChange.mock.calls[4][0]).toBe('喜马拉雅山'))
    await waitFor(() => expect(onChange.mock.calls[5][0]).toBe(''))
    await waitFor(() => expect(onSearch.mock.calls[0][0]).toBe(''))
  })

  test('外部输入值，搜索', async () => {
    const onSearch = jest.fn()
    const onChange = jest.fn()
    const screen = render(<Search value="100" openInput onSearch={onSearch} onChange={onChange} />)
    const input = screen.getByTestId('input')
    UserEvent.type(input, '喜马拉雅山')
    UserEvent.click(screen.getByTestId('search-btn'))
    await waitFor(() => expect(onChange).toHaveBeenCalledTimes(5))
    await waitFor(() => expect(onSearch).toHaveBeenCalledTimes(1))
    await waitFor(() => expect(onChange.mock.calls[0][0]).toBe('100喜'))
    await waitFor(() => expect(onChange.mock.calls[1][0]).toBe('100马'))
    await waitFor(() => expect(onChange.mock.calls[2][0]).toBe('100拉'))
    await waitFor(() => expect(onChange.mock.calls[3][0]).toBe('100雅'))
    await waitFor(() => expect(onChange.mock.calls[4][0]).toBe('100山'))
    await waitFor(() => expect(onSearch.mock.calls[0][0]).toBe(100))
  })

  test('设置提示语', async () => {
    const screen = render(<Search openInput placeholder="请输入关键词" />)
    screen.getByText('请输入关键词')
    const input = screen.getByTestId('input')
    UserEvent.type(input, '喜马拉雅山')
    await waitFor(async () => expect(await screen.queryByText('请输入关键词')).toBeNull())
  })

  test('设置提示语列表', async () => {
    const screen = render(<Search openInput placeholder={[
      '常青藤爸爸',
      '向日葵妈妈',
      '请输入关键字搜索'
    ]} />)
    screen.getByText('常青藤爸爸')
    screen.getByText('向日葵妈妈')
    screen.getByText('请输入关键字搜索')
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('设置提示列表 输入内容', async () => {
    const screen = render(<Search openInput placeholder={[
      '常青藤爸爸',
      '向日葵妈妈',
      '请输入关键字搜索'
    ]} />)
    screen.getByText('常青藤爸爸')
    screen.getByText('向日葵妈妈')
    screen.getByText('请输入关键字搜索')
    const input = screen.getByTestId('input')
    UserEvent.type(input, '喜马拉雅山')
    await waitFor(() => expect(screen.queryByText('常青藤爸爸')).toBeNull())
    await waitFor(() => expect(screen.queryByText('向日葵妈妈')).toBeNull())
    await waitFor(() => expect(screen.queryByText('请输入关键字搜索')).toBeNull())
  })

  test('设置提示列表 失去焦', async () => {
    const screen = render(<Search openInput placeholder={[
      '常青藤爸爸',
      '向日葵妈妈',
      '请输入关键字搜索'
    ]} />)
    screen.getByText('常青藤爸爸')
    screen.getByText('向日葵妈妈')
    screen.getByText('请输入关键字搜索')
    const input = screen.getByTestId('input')
    fireEvent.focus(input)
    await waitFor(() => expect(screen.queryByText('常青藤爸爸')).toBeNull())
    await waitFor(() => expect(screen.queryByText('向日葵妈妈')).toBeNull())
    await waitFor(() => expect(screen.queryByText('请输入关键字搜索')).toBeNull())
    fireEvent.blur(input)
    await waitFor(() => expect(screen.queryByText('常青藤爸爸')).not.toBeNull())
    await waitFor(() => expect(screen.queryByText('向日葵妈妈')).not.toBeNull())
    await waitFor(() => expect(screen.queryByText('请输入关键字搜索')).not.toBeNull())
  })

  test('点击事件', async () => {
    const onBack = jest.fn()
    const onClick = jest.fn()
    const onSearch = jest.fn()
    const screen = render(<Search back onBack={onBack} onClick={onClick} onSearch={onSearch} />)
    fireEvent.click(screen.getByTestId('search-box'))
    expect(onClick).toHaveBeenCalledTimes(1)
    expect(onBack).toHaveBeenCalledTimes(0)
    expect(onSearch).toHaveBeenCalledTimes(0)
  })

  test('点击事件 同时开启输入', async () => {
    const onBack = jest.fn()
    const onClick = jest.fn()
    const onSearch = jest.fn()
    const screen = render(<Search back openInput onBack={onBack} onClick={onClick} onSearch={onSearch} />)
    fireEvent.click(screen.getByTestId('search-box'))
    expect(onClick).toHaveBeenCalledTimes(1)
    expect(onBack).toHaveBeenCalledTimes(0)
    expect(onSearch).toHaveBeenCalledTimes(0)
  })
})
