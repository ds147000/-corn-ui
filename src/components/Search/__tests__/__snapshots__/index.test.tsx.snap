// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search 基础快照 1`] = `
<div>
  <form>
    <div
      class="xrk-search xrk-f xrk-ac xrk-search-undefined"
    >
      <div
        class="xrk-search-input-box xrk-f xrk-ac"
        data-testid="search-box"
      >
        <i
          class="xrk-search-icon xrk-icon xrk-i-search"
        />
        <div
          class="xrk-search-input"
        />
        <div />
        <div />
        <button
          class="xrk-btn xrk-if xrk-ac xrk-jc xrk-btn-middle xrk-btn-primary xrk-search-btn"
          data-testid="search-btn"
          type="submit"
        >
          搜索
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`Search 基础快照 light 1`] = `
<div>
  <form>
    <div
      class="xrk-search xrk-f xrk-ac xrk-search-light"
    >
      <div
        class="xrk-search-input-box xrk-f xrk-ac"
        data-testid="search-box"
      >
        <i
          class="xrk-search-icon xrk-icon xrk-i-search"
        />
        <div
          class="xrk-search-input"
        />
        <div />
        <div />
        <button
          class="xrk-btn xrk-if xrk-ac xrk-jc xrk-btn-middle xrk-btn-primary xrk-search-btn"
          data-testid="search-btn"
          type="submit"
        >
          搜索
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`Search 插入尾部内容 1`] = `
<div>
  <form>
    <div
      class="xrk-search xrk-f xrk-ac xrk-search-undefined"
    >
      <div
        class="xrk-search-input-box xrk-f xrk-ac"
        data-testid="search-box"
      >
        <i
          class="xrk-search-icon xrk-icon xrk-i-search"
        />
        <div
          class="xrk-search-input"
        />
        <div />
        <div />
        <button
          class="xrk-btn xrk-if xrk-ac xrk-jc xrk-btn-middle xrk-btn-primary xrk-search-btn"
          data-testid="search-btn"
          type="submit"
        >
          搜索
        </button>
      </div>
      <div
        class="xrk-search-suffix xrk-f xkr-ac"
      >
        <div>
          打卡
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`Search 设置提示语列表 1`] = `
<div>
  <form>
    <div
      class="xrk-search xrk-f xrk-ac xrk-search-undefined"
    >
      <div
        class="xrk-search-input-box xrk-f xrk-ac"
        data-testid="search-box"
      >
        <i
          class="xrk-search-icon xrk-icon xrk-i-search"
        />
        <input
          autocomplete="off"
          class="weui-input xrk-search-input xrk-input"
          confirmtype="search"
          data-testid="input"
          name="value"
          type="search"
          value=""
        />
        <div>
          <div
            class="xrk-search-placeholder"
          >
            <div
              class="xrk-search-placeholder-list"
            >
              <div
                class="xrk-search-placeholder-item xrk-f xrk-ac"
              >
                常青藤爸爸
              </div>
              <div
                class="xrk-search-placeholder-item xrk-f xrk-ac"
              >
                @Corn
              </div>
              <div
                class="xrk-search-placeholder-item xrk-f xrk-ac"
              >
                请输入关键字搜索
              </div>
            </div>
          </div>
        </div>
        <div />
        <button
          class="xrk-btn xrk-if xrk-ac xrk-jc xrk-btn-middle xrk-btn-primary xrk-search-btn"
          data-testid="search-btn"
          type="submit"
        >
          搜索
        </button>
      </div>
    </div>
  </form>
</div>
`;
